examples/src/bin/gif_paintable.rs

```diff
@@ -0,0 +1,305 @@
+        let decoder = GifDecoder::new(read)?;
+        let frames = decoder
+            .into_frames()
+            .collect_frames()?
+            .into_iter()
+            .map(Frame::new)
+            .collect::<Vec<Frame>>();
+
+        self_.frames.replace(frames);
+        self.invalidate_contents();
+        glib::timeout_add_local(
+            std::time::Duration::from_millis(60),
+            glib::clone!(@weak self as paintable => move || {
+                paintable.setup_next_frame();
+                glib::Continue(true)
+            }),
+        );
+        Ok(())
+    }
+
+    fn setup_next_frame(&self) {
+        let self_ = imp::GifPaintable::from_instance(self);
+        let mut idx = self_.current_idx.get() + 1;
+        let frames = self_.frames.borrow();
+        if idx >= frames.len() {
+            idx = 0;
+        }
+        self_.current_idx.set(idx);
+
+        let frame = frames.get(idx).unwrap();
+
+        self_.width.set(frame.width);
+        self_.height.set(frame.height);
+        self_.next_frame.replace(Some(frame.texture.clone()));
+
+        self.invalidate_contents();
+    }
+}
+
+mod win {
+    use super::*;
+    use glib::subclass;
+    use std::cell::RefCell;
+
+    #[derive(Debug)]
+    pub struct Window {
+        pub paintable: GifPaintable,
+        pub file_chooser: RefCell<Option<gtk::FileChooserNative>>,
+    }
+
+    impl ObjectSubclass for Window {
+        const NAME: &'static str = "Window";
+        type Type = super::Window;
+        type ParentType = gtk::ApplicationWindow;
+        type Instance = subclass::simple::InstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
+
+        glib::object_subclass!();
+
+        fn new() -> Self {
+            Self {
+                paintable: GifPaintable::new(),
+                file_chooser: RefCell::new(None),
+            }
+        }
+    }
+
+    impl ObjectImpl for Window {
+        fn constructed(&self, obj: &Self::Type) {
+            obj.setup_widgets();
+            obj.setup_actions();
+            self.parent_constructed(obj);
+        }
+    }
+    impl WidgetImpl for Window {}
+    impl WindowImpl for Window {}
+    impl ApplicationWindowImpl for Window {}
+}
+
+glib::wrapper! {
+    pub struct Window(ObjectSubclass<win::Window>) @extends gtk::Widget, gtk::Window, gtk::ApplicationWindow, @implements gio::ActionGroup, gio::ActionMap;
+}
+
+impl Window {
+    pub fn new<P: glib::IsA<gtk::Application>>(app: &P) -> Self {
+        glib::Object::new(&[("application", app)]).expect("Failed to create Window")
+    }
+
+    fn open_file(&self, file: gio::File) -> Result<(), Box<dyn std::error::Error>> {
+        let self_ = win::Window::from_instance(self);
+        let (bytes, _) = file.load_contents(gio::NONE_CANCELLABLE)?;
+        self_.paintable.load_form_bytes(bytes)?;
+        Ok(())
+    }
+
+    fn open_file_dialog(&self) {
+        let self_ = win::Window::from_instance(self);
```

Comment:
Maybe don't allow creating a new file chooser if one is open already

Reply:
The file chooser is modal, so the user can't focus the parent window

Reply:
Worth an assertion here then